File: test-automation-core/src/main/java/com/nortal/test/core/plugin/FeatureSourcesHelper.java
package com.nortal.test.core.plugin;

import io.cucumber.messages.Messages;
import io.cucumber.plugin.event.DataTableArgument;
import io.cucumber.plugin.event.DocStringArgument;
import io.cucumber.plugin.event.PickleStepTestStep;
import io.cucumber.plugin.event.Result;
import io.cucumber.plugin.event.Status;
import io.cucumber.plugin.event.StepArgument;
import io.cucumber.plugin.event.TestCase;
import io.cucumber.plugin.event.TestCaseStarted;
import io.cucumber.plugin.event.TestSourceRead;
import io.cucumber.plugin.event.TestStep;
import io.cucumber.plugin.event.TestStepStarted;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

/**
 * This class helps with parsing cucumber feature files and creating json structure of the report.
 * Most of the methods here and in TestSourcesModel class are lifted from cucumber source as is.
 */
public class FeatureSourcesHelper {
    private static final String STEPS = "steps";
    private static final String KEYWORD = "keyword";
    private static final String DESCRIPTION = "description";

    private final TestSourcesModel testSources = new TestSourcesModel();

    public void addSources(TestSourceRead event) {
        testSources.addTestSourceReadEvent(event.getUri().toString(), event);
    }

    public String getFeatureName(String uri) {
        return testSources.getFeatureName(uri);
    }

    public boolean hasBackground(TestCaseStarted event) {
        return testSources.hasBackground(event.getTestCase().getUri().toString(), event.getTestCase().getLine());
    }

    public boolean isBackgroundStep(TestStepStarted step) {
        TestSourcesModel.AstNode astNode = testSources.getAstNode(step.getTestCase().getUri().toString(), ((PickleStepTestStep) step.getTestStep()).getStepLine());
        return TestSourcesModel.isBackgroundStep(astNode.parent);
    }

    public Map<String, Object> createFeatureMap(TestCase testCase) {
        Map<String, Object> featureMap = new HashMap<>();
        featureMap.put("uri", testCase.getUri());
        featureMap.put("elements", new ArrayList<Map<String, Object>>());
        Messages.GherkinDocument.Feature feature = testSources.getFeature(testCase.getUri().toString());
        if (feature != null) {
            featureMap.put(KEYWORD, feature.getKeyword());
            featureMap.put("name", feature.getName());
            featureMap.put(DESCRIPTION, feature.getDescription() != null ? feature.getDescription() : "");
            featureMap.put("line", feature.getLocation().getLine());
            featureMap.put("id", TestSourcesModel.convertToId(feature.getName()));
            featureMap.put("tags", transformTags(feature));
        }
        return featureMap;
    }

    private List<Map<String, Object>> transformTags(final Messages.GherkinDocument.Feature feature) {
        return feature
            .getTagsList()
            .stream()
            .map(tag -> Map.of("name", tag.getName(), "type", "Tag", "location",
                Map.of("line", tag.getLocation().getLine(), "column", tag.getLocation().getColumn())
            ))
            .collect(Collectors.toList());
    }

    public Map<String, Object> createMatchMap(TestStep step, Result result) {
        Map<String, Object> matchMap = new HashMap<>();
        if (step instanceof PickleStepTestStep) {
            PickleStepTestStep testStep = (PickleStepTestStep) step;
            if (!testStep.getDefinitionArgument().isEmpty()) {
                List<Map<String, Object>> argumentList = new ArrayList<>();
                for (io.cucumber.plugin.event.Argument argument : testStep.getDefinitionArgument()) {
                    Map<String, Object> argumentMap = new HashMap<>();
                    if (argument.getValue() != null) {
                        argumentMap.put("val", argument.getValue());
                        argumentMap.put("offset", argument.getStart());
                    }
                    argumentList.add(argumentMap);
                }
                matchMap.put("arguments", argumentList);
            }
        }
        if (!result.getStatus().is(Status.UNDEFINED)) {
            matchMap.put("location", step.getCodeLocation());
        }
        return matchMap;
    }

    public Map<String, Object> createResultMap(Result result) {
        Map<String, Object> resultMap = new HashMap<>();
        resultMap.put("status", result.getStatus().toString());
        if (result.getError() != null) {
            resultMap.put("error_message", result.getError().getMessage());
        }
        if (result.getDuration() != null && !result.getDuration().isZero()) {
            resultMap.put("duration", result.getDuration().toNanos());
        }
        return resultMap;
    }

    public Map<String, Object> createTestCase(TestCaseStarted event) {
        TestCase testCase = event.getTestCase();
        Map<String, Object> testCaseMap = new HashMap<>();
        testCaseMap.put("name", testCase.getName());
        testCaseMap.put("line", testCase.getLine());
        testCaseMap.put("type", "scenario");
        TestSourcesModel.AstNode astNode = testSources.getAstNode(testCase.getUri().toString(), testCase.getLine());
        if (astNode != null) {
            testCaseMap.put("id", TestSourcesModel.calculateId(astNode));
            Messages.GherkinDocument.Feature.Scenario scenarioDefinition =
                TestSourcesModel.getScenarioDefinition(astNode);
            testCaseMap.put(KEYWORD, scenarioDefinition.getKeyword());
            testCaseMap.put(DESCRIPTION, scenarioDefinition.getDescription() != null
                                         ? scenarioDefinition.getDescription()
                                         : "");
        }
        testCaseMap.put(STEPS, new ArrayList<Map<String, Object>>());
        if (!testCase.getTags().isEmpty()) {
            List<Map<String, Object>> tagList = new ArrayList<>();
            for (String tag : testCase.getTags()) {
                Map<String, Object> tagMap = new HashMap<>();
                tagMap.put("name", tag);
                tagList.add(tagMap);
            }
            testCaseMap.put("tags", tagList);
        }
        return testCaseMap;
    }

    public Map<String, Object> createBackground(TestCaseStarted event) {
        TestCase testCase = event.getTestCase();
        TestSourcesModel.AstNode astNode = testSources.getAstNode(testCase.getUri().toString(), testCase.getLine());
        if (astNode != null) {
            Messages.GherkinDocument.Feature.Background background = TestSourcesModel.getBackgroundForTestCase(astNode);
            Map<String, Object> testCaseMap = new HashMap<>();
            testCaseMap.put("name", background.getName());
            testCaseMap.put("line", background.getLocation().getLine());
            testCaseMap.put("type", "background");
            testCaseMap.put(KEYWORD, background.getKeyword());
            testCaseMap.put(DESCRIPTION, background.getDescription() != null ? background.getDescription() : "");
            testCaseMap.put(STEPS, new ArrayList<Map<String, Object>>());
            return testCaseMap;
        }
        return null;
    }

    public Map<String, Object> createTestStep(TestStepStarted testStepEvent) {
        io.cucumber.plugin.event.Step testStep = ((PickleStepTestStep) testStepEvent.getTestStep()).getStep();
        Map<String, Object> stepMap = new HashMap<>();
        stepMap.put("name", testStep.getText());
        stepMap.put("line", testStep.getLine());
        TestSourcesModel.AstNode astNode = testSources.getAstNode(testStepEvent.getTestCase().getUri().toString(), testStep.getLine());
        if (testStep.getArgument() != null) {
            StepArgument argument = testStep.getArgument();
            if (argument instanceof DocStringArgument) {
                stepMap.put("doc_string", createDocStringMap(argument));
            } else if (argument instanceof DataTableArgument) {
                stepMap.put("rows", createDataTableList(argument));
            }
        }
        if (astNode != null) {
            Messages.GherkinDocument.Feature.Step step = (Messages.GherkinDocument.Feature.Step) astNode.node;
            stepMap.put(KEYWORD, step.getKeyword());
        }

        return stepMap;
    }

    private Map<String, Object> createDocStringMap(StepArgument argument) {
        Map<String, Object> docStringMap = new HashMap<>();
        DocStringArgument docString = (DocStringArgument) argument;
        docStringMap.put("value", docString.getContent());
        docStringMap.put("line", docString.getLine());
        docStringMap.put("content_type", docString.getContentType());
        return docStringMap;
    }

    private List<Map<String, List<String>>> createDataTableList(StepArgument argument) {
        List<Map<String, List<String>>> rowList = new ArrayList<>();
        for (List<String> row : ((DataTableArgument)argument).cells()) {
            Map<String, List<String>> rowMap = new HashMap<>();
            rowMap.put("cells", new ArrayList<>(row));
            rowList.add(rowMap);
        }
        return rowList;
    }

}

File: test-automation-core-old/src/main/java/com/nortal/test/core/plugin/FeatureSourcesHelper.kt
package com.nortal.test.core.plugin

import com.nortal.test.core.plugin.TestSourcesModel.addTestSourceReadEvent
import com.nortal.test.core.plugin.TestSourcesModel.getFeatureName
import com.nortal.test.core.plugin.TestSourcesModel.hasBackground
import com.nortal.test.core.plugin.TestSourcesModel.getAstNode
import com.nortal.test.core.plugin.TestSourcesModel.Companion.isBackgroundStep
import com.nortal.test.core.plugin.TestSourcesModel.getFeature
import com.nortal.test.core.plugin.TestSourcesModel.Companion.convertToId
import com.nortal.test.core.plugin.TestSourcesModel.Companion.calculateId
import com.nortal.test.core.plugin.TestSourcesModel.Companion.getScenarioDefinition
import com.nortal.test.core.plugin.TestSourcesModel.Companion.getBackgroundForTestCase
import java.util.HashMap
import io.cucumber.messages.Messages.GherkinDocument
import io.cucumber.messages.Messages.GherkinDocument.Feature.Background
import io.cucumber.plugin.event.*
import java.util.ArrayList
import java.util.stream.Collectors

/**
 * This class helps with parsing cucumber feature files and creating json structure of the report.
 * Most of the methods here and in TestSourcesModel class are lifted from cucumber source as is.
 */
class FeatureSourcesHelper {
    private val testSources = TestSourcesModel()
    fun addSources(event: TestSourceRead) {
        testSources.addTestSourceReadEvent(event.uri.toString(), event)
    }

    fun getFeatureName(uri: String?): String {
        return testSources.getFeatureName(uri!!)
    }

    fun hasBackground(event: TestCaseStarted): Boolean {
        return testSources.hasBackground(event.testCase.uri.toString(), event.testCase.line)
    }

    fun isBackgroundStep(step: TestStepStarted): Boolean {
        val astNode = testSources.getAstNode(step.testCase.uri.toString(), (step.testStep as PickleStepTestStep).stepLine)
        return isBackgroundStep(astNode!!.parent!!)
    }

    fun createFeatureMap(testCase: TestCase): Map<String, Any> {
        val featureMap: MutableMap<String, Any> = HashMap()
        featureMap["uri"] = testCase.uri
        featureMap["elements"] = ArrayList<Map<String, Any>>()
        val feature = testSources.getFeature(testCase.uri.toString())
        if (feature != null) {
            featureMap[KEYWORD] = feature.keyword
            featureMap["name"] = feature.name
            featureMap[DESCRIPTION] = if (feature.description != null) feature.description else ""
            featureMap["line"] = feature.location.line
            featureMap["id"] = convertToId(feature.name)
            featureMap["tags"] = transformTags(feature)
        }
        return featureMap
    }

    private fun transformTags(feature: GherkinDocument.Feature): List<Map<String, Any>> {
        return feature
            .tagsList
            .stream()
            .map { tag: GherkinDocument.Feature.Tag ->
                java.util.Map.of(
                    "name", tag.name, "type", "Tag", "location",
                    java.util.Map.of("line", tag.location.line, "column", tag.location.column)
                )
            }
            .collect(Collectors.toList())
    }

    fun createMatchMap(step: TestStep, result: Result): Map<String, Any> {
        val matchMap: MutableMap<String, Any> = HashMap()
        if (step is PickleStepTestStep) {
            val testStep = step
            if (!testStep.definitionArgument.isEmpty()) {
                val argumentList: MutableList<Map<String, Any>> = ArrayList()
                for (argument in testStep.definitionArgument) {
                    val argumentMap: MutableMap<String, Any> = HashMap()
                    if (argument.value != null) {
                        argumentMap["val"] = argument.value
                        argumentMap["offset"] = argument.start
                    }
                    argumentList.add(argumentMap)
                }
                matchMap["arguments"] = argumentList
            }
        }
        if (!result.status.`is`(Status.UNDEFINED)) {
            matchMap["location"] = step.codeLocation
        }
        return matchMap
    }

    fun createResultMap(result: Result): Map<String, Any?> {
        val resultMap: MutableMap<String, Any?> = HashMap()
        resultMap["status"] = result.status.toString()
        if (result.error != null) {
            resultMap["error_message"] = result.error.message
        }
        if (result.duration != null && !result.duration.isZero) {
            resultMap["duration"] = result.duration.toNanos()
        }
        return resultMap
    }

    fun createTestCase(event: TestCaseStarted): Map<String, Any> {
        val testCase = event.testCase
        val testCaseMap: MutableMap<String, Any> = HashMap()
        testCaseMap["name"] = testCase.name
        testCaseMap["line"] = testCase.line
        testCaseMap["type"] = "scenario"
        val astNode = testSources.getAstNode(testCase.uri.toString(), testCase.line)
        if (astNode != null) {
            testCaseMap["id"] = calculateId(astNode)
            val scenarioDefinition = getScenarioDefinition(astNode)
            testCaseMap[KEYWORD] = scenarioDefinition.keyword
            testCaseMap[DESCRIPTION] = if (scenarioDefinition.description != null) scenarioDefinition.description else ""
        }
        testCaseMap[STEPS] = ArrayList<Map<String, Any>>()
        if (!testCase.tags.isEmpty()) {
            val tagList: MutableList<Map<String, Any>> = ArrayList()
            for (tag in testCase.tags) {
                val tagMap: MutableMap<String, Any> = HashMap()
                tagMap["name"] = tag
                tagList.add(tagMap)
            }
            testCaseMap["tags"] = tagList
        }
        return testCaseMap
    }

    fun createBackground(event: TestCaseStarted): Map<String, Any>? {
        val testCase = event.testCase
        val astNode = testSources.getAstNode(testCase.uri.toString(), testCase.line)
        if (astNode != null) {
            val background = getBackgroundForTestCase(astNode)
            val testCaseMap: MutableMap<String, Any> = HashMap()
            testCaseMap["name"] = background!!.name
            testCaseMap["line"] = background.location.line
            testCaseMap["type"] = "background"
            testCaseMap[KEYWORD] = background.keyword
            testCaseMap[DESCRIPTION] = if (background.description != null) background.description else ""
            testCaseMap[STEPS] = ArrayList<Map<String, Any>>()
            return testCaseMap
        }
        return null
    }

    fun createTestStep(testStepEvent: TestStepStarted): Map<String, Any> {
        val testStep = (testStepEvent.testStep as PickleStepTestStep).step
        val stepMap: MutableMap<String, Any> = HashMap()
        stepMap["name"] = testStep.text
        stepMap["line"] = testStep.line
        val astNode = testSources.getAstNode(testStepEvent.testCase.uri.toString(), testStep.line)
        if (testStep.argument != null) {
            val argument = testStep.argument
            if (argument is DocStringArgument) {
                stepMap["doc_string"] = createDocStringMap(argument)
            } else if (argument is DataTableArgument) {
                stepMap["rows"] = createDataTableList(argument)
            }
        }
        if (astNode != null) {
            val step = astNode.node as GherkinDocument.Feature.Step
            stepMap[KEYWORD] = step.keyword
        }
        return stepMap
    }

    private fun createDocStringMap(argument: StepArgument): Map<String, Any> {
        val docStringMap: MutableMap<String, Any> = HashMap()
        val docString = argument as DocStringArgument
        docStringMap["value"] = docString.content
        docStringMap["line"] = docString.line
        docStringMap["content_type"] = docString.contentType
        return docStringMap
    }

    private fun createDataTableList(argument: StepArgument): List<Map<String, List<String>>> {
        val rowList: MutableList<Map<String, List<String>>> = ArrayList()
        for (row in (argument as DataTableArgument).cells()) {
            val rowMap: MutableMap<String, List<String>> = HashMap()
            rowMap["cells"] = ArrayList(row)
            rowList.add(rowMap)
        }
        return rowList
    }

    companion object {
        private const val STEPS = "steps"
        private const val KEYWORD = "keyword"
        private const val DESCRIPTION = "description"
    }
}